test/core/event_engine/posix/posix_endpoint_test.cc

```diff
@@ -0,0 +1,352 @@
+  bool IsZeroCopyEnabled() const { return is_zero_copy_enabled_; }
+
+ private:
+  std::string poller_;
+  bool is_zero_copy_enabled_;
+};
+
+std::string TestScenarioName(const ::testing::TestParamInfo<TestParam>& info) {
+  return absl::StrCat("poller_type_", info.param.Poller(),
+                      "_is_zero_copy_enabled_", info.param.IsZeroCopyEnabled());
+}
+
+// A helper class to drive the polling of Fds. It repeatedly calls the Work(..)
+// method on the poller to get pet pending events, then schedules another
+// parallel Work(..) instantiation and processes these pending events. This
+// continues until all Fds have orphaned themselves.
+class Worker : public grpc_core::DualRefCounted<Worker> {
+ public:
+  Worker(std::shared_ptr<EventEngine> engine, PosixEventPoller* poller)
+      : engine_(std::move(engine)), poller_(poller) {
+    WeakRef().release();
+  }
+  void Orphan() override { signal.Notify(); }
+  void Start() {
+    // Start executing Work(..).
+    engine_->Run([this]() { Work(); });
+  }
+
+  void Wait() {
+    signal.WaitForNotification();
+    WeakUnref();
+  }
+
+ private:
+  void Work() {
+    auto result = poller_->Work(24h, [this]() {
+      // Schedule next work instantiation immediately and take a Ref for
+      // the next instantiation.
+      Ref().release();
+      engine_->Run([this]() { Work(); });
+    });
+    ASSERT_TRUE(result == Poller::WorkResult::kOk ||
+                result == Poller::WorkResult::kKicked);
+    // Corresponds to the Ref taken for the current instantiation. If the
+    // result was Poller::WorkResult::kKicked, then the next work instantiation
+    // would not have been scheduled and the poll_again callback would have
+    // been deleted.
+    Unref();
+  }
+  std::shared_ptr<EventEngine> engine_;
+  PosixEventPoller* poller_;
+  grpc_core::Notification signal;
+};
+
+class PosixEndpointTest : public ::testing::TestWithParam<TestParam> {
+  void SetUp() override {
+    posix_ee_ = std::make_shared<PosixEventEngine>();
+    scheduler_ =
+        absl::make_unique<grpc_event_engine::posix_engine::TestScheduler>(
+            posix_ee_.get());
+    EXPECT_NE(scheduler_, nullptr);
+    GPR_GLOBAL_CONFIG_SET(grpc_poll_strategy, GetParam().Poller().c_str());
+    poller_ = GetDefaultPoller(scheduler_.get());
+    if (poller_ != nullptr) {
+      EXPECT_EQ(poller_->Name(), GetParam().Poller());
+    }
+  }
+
+  void TearDown() override {
+    if (poller_ != nullptr) {
+      poller_->Shutdown();
+    }
+    WaitForPendingTasks(std::move(posix_ee_));
+  }
+
+ public:
+  TestScheduler* Scheduler() { return scheduler_.get(); }
+
+  std::shared_ptr<EventEngine> GetPosixEE() { return posix_ee_; }
+
+  PosixEventPoller* PosixPoller() { return poller_; }
+
+ private:
+  PosixEventPoller* poller_;
+  std::unique_ptr<TestScheduler> scheduler_;
+  std::shared_ptr<EventEngine> posix_ee_;
+};
+
+TEST_P(PosixEndpointTest, ConnectExchangeBidiDataTransferTest) {
+  if (PosixPoller() == nullptr) {
```

Comment:
What is/are the valid scenario(s) in which this test should pass but the poller == nullptr? Should this be an assert?

Reply:
PosixPoller will be nullptr if the specified poller (epoll/poll) is not supported on this platform. Were trying both pollers epoll and poll in these tests. Some platforms like MacOS dont support epoll pollers.

Reply:
Doesn't the build system already enforce that epoll1 tests do not run on mac? The BUILD file shows that `posix_endpoint_test` is already configured with `uses_polling=True`, which means you'll have test targets like `posix_endpoint_test@poller=poll` and `posix_endpoint_test@poller=epoll1`. I'd suggest using the build system's tooling to enforce this rather than reinventing that logic in the test setup itself.

Reply:
The intent is to test the functionality of GetDefaultPoller in line 218 as well. It reads the poll_strategy environment variable and returns the appropriate poller. In the test setup, we override the poll_strategy environment variable and in line 220 we check whether the name of the poller returned by GetDefaultPoller matches the set environment variable.

Reply:
I simplified the testing logic per our offline discussion. But I'd prefer to have the nullptr check still to handle some corner cases. For e.g poller might be epoll, but the system may not support creating epoll_fds. In that case, GetDefaultPoller will return nullptr.